### PR TITLE
fix(redraw): avoid unnecessary redraws and glitches with floats+messages

### LINKED ISSUE #20106: bug(cmdheight=0): floating window above cmdline fails to redraw (using whichkey, cmp-cmdline)
### Neovim version (nvim -v)

v0.8.0-dev-2497-gfd1595514-dirty 

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

macOS 12.4

### Terminal name/version

Kitty 0.25.2

### $TERM environment variable

xterm-kitty

### Installation

homebrew

### How to reproduce the issue

1. Install `which-key.nvim` it is hard to reproduce the behaviour of overlapping windows on key press from the commandline without it.

2. run `git checkout 7ee35a7614e34e562fd3f815ad35bd6d7e456093` as the author has added in a workaround which no longer triggers this issue (seems to hide it whilst a fix is being awaited)

Author's workaround 👉🏿 https://github.com/folke/which-key.nvim/commit/c96627900191355e6788629bbf5239d7295221f0

```lua
set rtp+=~/src/which-key.nvim

vim.o.cmdheight = 0

vim.cmd([[
nnoremap <Leader>aa   <Cmd>echo "hoge"<CR>
nnoremap <Leader>ba   <Cmd>echo "hoge"<CR>
nnoremap <Leader>ca   <Cmd>echo "hoge"<CR>
nnoremap <Leader>da   <Cmd>echo "hoge"<CR>
nnoremap <Leader>ea   <Cmd>echo "hoge"<CR>
nnoremap <Leader>fg   <Cmd>echo "hoge"<CR>
]])

```

3. `nvim -u <minimal-config-above>`
4. Press `<leader>a` wait a moment
5. press `a` you should see both windows overlap

### Expected behavior

The second window replaces the first. This is an exact regression of #19184 and reproduces in the same way. The cause is likely related to https://github.com/neovim/neovim/pull/19816#issuecomment-1238284631 i.e. the difference in the redraw mechanism mentioned in that PR.

This issues stems from https://github.com/neovim/neovim/pull/19816#issuecomment-1238822565.

@Shougo there is no actual difference between the previous reproduction from #19184 I just didn't notice that the author added a workaround which hides this issue, so you need to check out a commit before the workaround. I have added it to reproduction instructions

### Actual behavior

https://user-images.githubusercontent.com/22454918/188838488-c3ace672-44c1-4de9-9438-6a045fb468da.mov

### PR DESCRIPTION
**Pull request by bfredl**
fixes https://github.com/neovim/neovim/issues/20106 and maybe others. will go through all recent cmdheight issues and write tests.

### CODE REVIEW

**Comment by Shougo** [comment]
I have tested it.

It fixes #20106 #20229 #20243

**Comment by Shougo** [comment]
It is better.  But it is not perfect.

It clears current command line like this.

https://github.com/neovim/neovim/pull/20233#issuecomment-1250084410

**Comment by zeertzjq** [comment]
> It clears current command line like this.

Does <ttps://github.com/vim/vim/commit/a653e53b1f4ba08de0dbcea06288cf0cc1c6e752> fix this?

**Comment by Shougo** [comment]
```diff
diff --git a/src/nvim/ex_docmd.c b/src/nvim/ex_docmd.c
index 52841b602..7d126882d 100644
--- a/src/nvim/ex_docmd.c
+++ b/src/nvim/ex_docmd.c
@@ -6077,6 +6077,11 @@ static void ex_redraw(exarg_T *eap)
   // No need to wait after an intentional redraw.
   need_wait_return = false;
 
+  // When invoked from a callback or autocmd the command line may be active.
+  if (State & MODE_CMDLINE) {
+    redrawcmdline();
+  }
+
   ui_flush();
 }
 ```

Unfortunately, no...

File: runtime/lua/man.lua [hunk 1]
```diff
@@ -23,13 +23,15 @@ local function man_system(cmd, silent)
   local done = false
   local exit_code
 
-  local handle = vim.loop.spawn(cmd[1], {
+  local handle
+  handle = vim.loop.spawn(cmd[1], {
```
Comment (justinmk) [inline runtime/lua/man.lua:27]: off-topic: man_system seems like something that would go in the stdlib.

**Pull request merged**